### ISSUE TITLE
Gap Between KLEE and Tracer-X Heap Modeling

### ISSUE THREAD

**Issue by domainexpert** [tracer-x/klee#247]
To keep track of pointers and memory bounds interpolation, Tracer-X
propagates the addresses of allocations (both heap and local
allocations) to pointers, such that each pointer is associated with
allocations. However, KLEE adopts a different model where each
allocation is stored in a binary tree structure such that there at
memory access, the validity is established by finding a valid range
on the structure.  Hence, the Tracer-X model is conservative in the
sense that if a pointer access is out of range of all its associated
allocations, this implies that the pointer access is invalid.  Note,
however, that Tracer-X leaves it to KLEE to detect memory errors.
KLEE's model, however, as long as a memory access falls within a valid
range, it can still be valid.  When the pointer is symbolic, KLEE
produces internal conditional branchings for each of its registered
memory region, which has a cascading structure. Each of the branch has
as its condition the (in-)validity constraint of the memory access
wrt. an allocation. When invalid access is assumed, the control
continues further to another branching that adds access (in-)validity
constraint of another region, and so on until all regions are
exhausted. Therefore, KLEE may discover a valid access when, under
Tracer-X model the access was invalid.

A problem here is that KLEE can decide that an access is valid when
Tracer-X cannot find a proof. In such a case, Tracer-X should
conservatively falls back to normal dependency computation instead of
memory bounds interpolation.

We first need an example to demonstrate the issue.

**Comment by domainexpert** [comment]
This is an example that demonstrates the issue. As at tracer-x/klee@c503ba2b2b7756ea99137fc82f06326dd9e1c19f, when the example is put in `basic` folder of `klee-examples` repo and executed (using DFS strategy), the assertion in `TxStateAddress::adjustOffsetBound` is violated due to KLEE deciding that the access is valid when Tracer-X is deciding the the access is invalid. It is currently resolved in tracer-x/klee#264 by weakening the assertion. It remains to be checked whether the weakening results in graceful falling back to normal dependency computation instead of bound interpolation.
```
/*
 * Simple regular expression matching.
 *
 * From:
 *   The Practice of Programming
 *   Brian W. Kernighan, Rob Pike
 *
 * This code with KLEE harness is obtained from the following
 * page of the KLEE tutorial on December 2015:
 *   http://klee.github.io/resources/Regexp.c.html
 */
#ifdef LLBMC
#include <llbmc.h>
#else
#include <klee/klee.h>
#endif

char _bound[6];
char *wcet;
void tracerx_check(char *p) { printf("Timing of Path:%d\n",(int)p-(int)(&(_bound))); *p;}

static int matchhere(char*,char*);

static int matchstar(int c, char *re, char *text) {
  do {
    if (matchhere(re, text)){
      wcet++;	
      return 1;
    }
  } while (*text != '\0' && (*text++ == c || c== '.'));
  return 0;
}

static int matchhere(char *re, char *text) {
  if (re[0] == '\0'){
     wcet++;
     return 0;
  }
  if (re[1] == '*'){
    return matchstar(re[0], re+2, text);
    wcet++;
  }
  if (re[0] == '$' && re[1]=='\0'){
    wcet++;
    return *text == '\0';
  }
  if (*text!='\0' && (re[0]=='.' || re[0]==*text)){
    wcet++;
    return matchhere(re+1, text+1);
  }
  return 0;
}

int match(char *re, char *text) {
  if (re[0] == '^'){
    wcet++;
    return matchhere(re+1, text);
  }
  do {
    if (matchhere(re, text)){
      wcet++;
      return 1;
    }
  } while (*text++ != '\0');
  return 0;
}

/*
 * Harness for testing with KLEE.
 */

// The size of the buffer to test with.
#define SIZE 7

int main() {
  wcet=_bound;
  // The input regular expression.
  char re[SIZE];
  wcet++;

// Make the input symbolic.
#ifdef LLBMC
  for (int i = 0; i < SIZE; ++i) {
    re[i] = __llbmc_nondef_char();
  }
#else
  klee_make_symbolic(re, sizeof re, "re");
  klee_assume(re[SIZE-1] == '\0');
#endif

  // Try to match against a constant string "hello".
  match(re, "hello");
  tracerx_check(wcet);
  return 0;
}
```

**Comment by domainexpert** [comment]
#266 makes sure that we fall back to normal interpolation (marking exact value) when memory bound check fails.
